### DISCUSSION TITLE
New CLI to convert SVGs to Quasar format?

### DISCUSSION THREAD

**Discussion by hawkeye64** [quasarframework/quasar#8683] [Ideas / Proposals]
This is an issue for discussion with the community.

The proposal for a CLI to take existing SVGs and converts them to [Quasar SVG format](https://next.quasar.dev/vue-components/icon#svg-icon-format).

First of all, it'd be nice to know how many people want this feature in the first place. A few people, in the past year, have asked if there is a tool for making this conversion. How big is the need? Is it convenience only? Usually, only a dozen (or less) icons are used from a library, unless you have very specific needs.

One issue to consider is that a lot of SVG libraries are made from tools that generalize the SVG to the point where it's added things that can't be resolved by a CLI. Sometimes these can be cleaned up enough by hand. But, the issue repeats itself each time there is a new release. So, how do we handle such failure? What is expected from the User? Can we be graceful about it, or be more frank that it can't be converted without manual intervention?

Please feel free to add your comments and/or questions below.

**Comment by eduardodgarciac** [comment]
I made an script to transform [https://tabler-icons.io/](url) to quasar svg format but works for others too.

**I share it here!!**

```js
const _ = require('lodash');
const fs = require('fs-extra');
const path = require('path');
const svgSon = require('svgson');
const toPath = require('element-to-path');

const svgStyleAttributes = ['stroke', 'stroke-width', 'stroke-linecap', 'stroke-linejoin', 'fill'];
const svgTransformAttributes = ['transform'];

let content = '';
const iconPackName = process.argv[2];
const iconPackPath = path.join(__dirname, `packs/${iconPackName}`);

(async () => {
    const files = fs.readdirSync(iconPackPath);
    for (const file of files) {
        if (!file.endsWith('.svg')) continue;

        // Prefix the name with 'svg'
        const fileName = `svg-${file.replace('.svg', '')}`;
        const filePath = path.join(iconPackPath, file);
        // Get svg string content
        const svgStr = fs.readFileSync(filePath, { encoding: 'utf-8' });
        // Parse svg string to object
        const svgObj = svgSon.parseSync(svgStr);
        // Extract viewBox and styles
        const svgViewBox = svgObj.attributes.viewBox;
        const svgStyle = jsonToStyle(svgObj.attributes, svgStyleAttributes);
        // Iterate all child elements to convert svg to quasar svg format
        const svgPath = svgObj.children.map(obj => {
            // Extract styles and transformations
            const styles = jsonToStyle(obj.attributes, svgStyleAttributes) || svgStyle;
            const transformations = jsonToStyle(obj.attributes, svgTransformAttributes)
            // Convert element to path
            const path = toPath(obj).replace(/[^A-Za-z0-9.-]/gm, ' ').replace(/\s\s+/g, ' ');
            // Join path + styles + transformations with quasar format '@'
            return _.trim([path, styles, transformations].join('@@'), '@');
        }).join('&&');

        // Add icon to file
        const name = _.camelCase(fileName);
        const value = _.trim(`${svgPath}|${svgViewBox}`, '|');
        content += `export const ${name} = '${value}'\n`;
    }

    // Write final file with the iconPackName
    fs.writeFileSync(path.join(iconPackPath, `../${iconPackName}.js`), content, { encoding: 'utf-8' });
})()

function jsonToStyle (obj, attrs) {
    return Object.entries(_.pick(obj, attrs)).map(([k, v]) => `${k}:${v}`).join(';');
}
```

**Comment by hawkeye64** [comment]
@eduardodgarciac That's great! I :heart: community participation.

**Comment by metalsadman** [comment]
There's also https://github.com/lazaroofarrill/svg-stringify.

**Comment by hawkeye64** [comment]
In case anyone sees this thread, I ended up creating a new package here: https://github.com/hawkeye64/quasar-extras-svg-icons

**Reply by ibrainventures** [reply]
@hawkeye64 

hi,
whats the recommend way to convert all fontawesome pro 6.1.1 (and later) to the svg quasar format?
I have a yearly subscription / licence and need only the svg export. (not css - only svg).

**Reply by hawkeye64** [reply]
Look into @quasar/extras, the build folder, and then fontawesome-v6. Copy it to fontawesome-pro-v6.js add it to the build/index.js. Make a fontawesome-pro-v6 folder for the output. You will need to modify your fontawesome-pro-v6.js file. Update variables `packageName`, `distName`, `iconsetName`, `distFolder` and add to `iconTypes` (thin, light, duotone, etc).
Then `yarn build`
Copy the fontawesome-pro-v6 where you need it. Every time they release something new, update your copy of extras.

**Reply by ssuess** [reply]
@hawkeye64 it seems this process has changed since you wrote this? I am trying to do the same (but for v5) and I don't see any "build" folder under @quasar/extras, only the names of icon sets. Is there a new procedure? Thanks!

**Reply by yusufkandemir** [reply]
@ssuess https://github.com/quasarframework/quasar/tree/dev/extras/build
